### PR TITLE
Optimize PRBAC SQL queries

### PR DESCRIPTION
**Pull request by millerdev**
Review commits individually. Depends on https://github.com/dimagi/django-prbac/pull/14

Some things to consider:
- All roles and grants are stored in memory. However, the tables are pretty small. Here are the numbers from prod:
  - django_prbac_grant - 256 rows
  - django_prbac_role - 91 rows
  - django_prbac_userrole - 30 rows
  
  I think keeping them in memory is a reasonable tradeoff for the reduction it DB hits.
- This optimization focuses on the case where "role assignments" are not used. I checked, and none of our roles or grants are using assignments. However, things should continue to work if we start using assignments, but more optimization may be necessary.

Page used for testing http://localhost:8080/a/test/dashboard/project/
Profiling numbers come from django toolbar

Before:
- Cold cache  
  CPU Time: 15335.91 ms (maybe not accurate?)  
  212 SQL queries in 196.54ms
- Hot cache  
  CPU Time: 5509.83 ms  
  47 SQL queries in 44.86ms

After:
- Cold cache  
  CPU Time: 6499.19 ms  
  6 SQL queries in 11.87ms
- Hot cache  
  CPU Time: 4085.39 ms  
  4 SQL queries in 8.66 ms

@esoergel @biyeun cc @sravfeyn

### CODE REVIEW

File: corehq/apps/users/views/__init__.py [hunk 1]
```diff
@@ -494,11 +494,8 @@ def user_roles(self):
 
     @property
     def can_edit_roles(self):
-        try:
-            ensure_request_has_privilege(self.request, privileges.ROLE_BASED_ACCESS)
-        except PermissionDenied:
-            return False
-        return self.couch_user.is_domain_admin
+        return has_privilege(self.request, privileges.ROLE_BASED_ACCESS) \
+                and self.couch_user.is_domain_admin
```
Comment (dimagimon) [inline corehq/apps/users/views/__init__.py]: E126 continuation line over-indented for hanging indent

File: corehq/apps/users/views/__init__.py [hunk 2]
```diff
@@ -570,11 +567,8 @@ def user_roles(self):
...
```
Comment (dimagimon) [inline corehq/apps/users/views/__init__.py]: E126 continuation line over-indented for hanging indent

File: corehq/apps/accounting/models.py [hunk 3]
```diff
@@ -1332,7 +1332,8 @@ def set_billing_account_entry_point(self):
     def _get_plan_by_subscriber(cls, subscriber):
         active_subscriptions = cls.objects\
             .filter(subscriber=subscriber, is_active=True)\
-            .order_by('-date_created')[:2]
+            .order_by('-date_created')[:2]\
+            .select_related('plan_version__role')
```
Comment (millerdev) [inline corehq/apps/accounting/models.py]: This eliminates two queries, and instead uses INNER JOINs to fetch the related objects. Maybe I should trace the call sites of this function to see if a parameter to make this optional would make sense (for example, if some call sites are not accessing `plan_version` and/or `plan_version.role`).

**Comment by czue** [comment]
I think you need to add the submodule update

**Comment by czue** [comment]
just trying to wrap my head around the caching implications. in the following scenario:

1. I don't have permission to use cloudcare
2. I change my subscription to one that has permission
3. I load cloudcare

Does caching kick in at all? I think it doesn't, and the scenario it would, would be if we decided the enterprise plan should no longer have access to cloudcare. Just want to make sure that's right. If it is the first case I would advocate for clearing them when plans change.

File: corehq/apps/accounting/utils.py [hunk 4]
```diff
@@ -116,10 +116,9 @@ def domain_has_privilege(domain, privilege_slug, **assignment):
     from corehq.apps.accounting.models import Subscription
     try:
         plan_version = Subscription.get_subscribed_plan_by_domain(domain)[0]
```
Comment (sravfeyn) [inline corehq/apps/accounting/utils.py]: Since there will be lot of different permission-checks, could `Subscription.get_subscribed_plan_by_domain(domain)[0]` be cached at the request level and passed here, instead of fetching each time? (Does this sound correct)

**Comment by millerdev** [comment]
@czue wrote:
> 1. I don't have permission to use cloudcare
> 2. I change my subscription to one that has permission
> 3. I load cloudcare
> 
> Does caching kick in at all? I think it doesn't

That's correct. The level of caching done in this PR will not affect that case since [a fresh copy of the user's role is looked up on each request](https://github.com/dimagi/commcare-hq/blob/abcd219fe4d88e974e971220b7da244bc98ec0d1/corehq/apps/domain/middleware.py#L36-L37). The caching part comes in when calculating if the user's role has permission to access cloudcare. At that point the user's role is [swapped with a cached copy](https://github.com/dimagi/django-prbac/blob/3b987dc0ac82708bb3a6e0e87e136c4ada78c930/django_prbac/models.py#L164)

> the scenario [this new caching] would [affect], would be if we decided the enterprise plan should no longer have access to cloudcare.

That's correct. If we changed enterprise to no longer have access to cloudcare, then people on the enterprise plan would have, at a maximum, the length of the cache duration (1 minute) to use cloudcare before they lost that privilege.

File: corehq/apps/accounting/utils.py (hunk 4 above)
Comment (millerdev) [inline corehq/apps/accounting/utils.py]: That's a good point. The [plan version is cached on the request](https://github.com/dimagi/commcare-hq/blob/abcd219fe4d88e974e971220b7da244bc98ec0d1/corehq/apps/domain/middleware.py#L38). Unfortunately it looks like this function does not have easy access to the request. Maybe we should change that? I'd prefer to do that in a separate PR.

Comment (sravfeyn) [inline corehq/apps/accounting/utils.py]: Sounds good

**Comment by sravfeyn** [comment]
test fail looks real

File: corehq/apps/users/views/__init__.py (hunk 2 above)
Comment (dimagimon) [inline corehq/apps/users/views/__init__.py]: E126 continuation line over-indented for hanging indent

**Comment by esoergel** [comment]
Test failure looks real:
```
======================================================================
ERROR: testGetCasesCaching (corehq.apps.cloudcare.tests.test_api.CaseAPITest)
----------------------------------------------------------------------

Traceback (most recent call last):
  File "/home/travis/build/dimagi/commcare-hq/corehq/apps/cloudcare/tests/test_api.py", line 244, in testGetCasesCaching
    'use_cache': 'true',
  File "/home/travis/virtualenv/local/lib/python2.7/site-packages/django/test/client.py", line 473, in get
    response = super(Client, self).get(path, data=data, **extra)
  File "/home/travis/virtualenv/local/lib/python2.7/site-packages/django/test/client.py", line 280, in get
    return self.request(**r)
  File "/home/travis/virtualenv/local/lib/python2.7/site-packages/django/test/client.py", line 444, in request
    six.reraise(*exc_info)
  File "/home/travis/virtualenv/local/lib/python2.7/site-packages/django/core/handlers/base.py", line 105, in get_response
    response = middleware_method(request, callback, callback_args, callback_kwargs)
  File "/home/travis/virtualenv/local/lib/python2.7/site-packages/django/utils/decorators.py", line 29, in _wrapper
    return bound_func(*args, **kwargs)
  File "/home/travis/virtualenv/local/lib/python2.7/site-packages/django/views/decorators/debug.py", line 75, in sensitive_post_parameters_wrapper
    return view(request, *args, **kwargs)
  File "/home/travis/virtualenv/local/lib/python2.7/site-packages/django/utils/decorators.py", line 25, in bound_func
    return func(self, *args2, **kwargs2)
  File "/home/travis/build/dimagi/commcare-hq/corehq/apps/domain/middleware.py", line 29, in process_view
    self.apply_prbac(request)
  File "/home/travis/build/dimagi/commcare-hq/corehq/apps/domain/middleware.py", line 44, in apply_prbac
    request.role = Role.get_privilege('community_plan_v0').role

AttributeError: 'NoneType' object has no attribute 'role'
```

**Comment by esoergel** [comment]
g2m once tests pass

**Comment by millerdev** [comment]
Let's wait to merge this (also wait to update the submodule) until Monday. Don't think it will cause issues, but would hate for something to go wrong since I have a pretty busy weekend with getting ready for the bike trip.

**Comment by esoergel** [comment]
sounds good

**Comment by millerdev** [comment]
clearly more work to be done here....

File: corehq/apps/hqadmin/management/commands/cchq_prbac_bootstrap.py [hunk 5]
```diff
@@ -90,6 +90,7 @@ def ensure_grant(self, grantee_slug, priv_slug, dry_run=False):
             grantee = Role.objects.get(slug=grantee_slug)
             priv = Role.objects.get(slug=priv_slug)
 
+            Role.get_cache().clear()
             if grantee.has_privilege(priv):
```
Comment (millerdev) [inline corehq/apps/hqadmin/management/commands/cchq_prbac_bootstrap.py]: Took me forever to track this down. I struggled with HQ tests because they're so slow and the console output is overwhelming and mostly useless. I was hoping to find a way to log SQL queries while running tests, but that proved fruitless until I went directly to postgres and enabled statement logging.

**Comment by millerdev** [comment]
This is discouraging. I have tests that are failing on my local machine, but I can't figure out why. They're even failing on a fresh checkout of master. Therefore it's been difficult for me to figure out which failures are real issues caused by changes on this branch.

**Comment by benrudolph** [comment]
Looks like they've passed on travis. What tests aren't passing locally? Just checking, you have `SOUTH_TESTS_MIGRATE = False`

**Comment by millerdev** [comment]
Finally passed! Yay!! I think that means this is good to merge, assuming you peeps are happy with it.

`SOUTH_TESTS_MIGRATE = False` is in settings.py, and I have not overridden it, so that seems right to me. Will troubleshoot the failing tests on other channels.

**Comment by millerdev** [comment]
@amsagoff do we need to do anything special to measure if this changes HQ performance when this is deployed (presumably later today)? Theoretically it should affect nearly every request to HQ.

**Comment by czue** [comment]
:tada:

**Comment by amsagoff** [comment]
amazing! We need to wait a bit to get statistically significant results. I'll be doing this with all the changes people made on performance day (or trying to - so hard to control for all the different changes happening!).
I added it to the list with the deployment date set optimistically to today: https://docs.google.com/spreadsheets/d/1hMJ0PtXfxGJNE_dqqhzyocHJUfl8-guX2cnKY3SlKZ4/edit#gid=931792981
